File: services/ai-service/src/jobs/inventoryCron.js
const cron = require('node-cron');
const { sendTelegramNotification, redisClient } = require('../services/telegramService');
const { chatWithTools } = require('../services/deepseekService');
const { buildSystemPrompt } = require('../services/systemPrompt');
const { buildSystemContext } = require('../services/contextBuilder');
const { TOOLS, executeTool } = require('../services/toolExecutor');

// Función que ejecuta el análisis de inventario de forma autónoma
async function runAutonomousInventoryCheck() {
  console.log('[Cron] Iniciando revisión autónoma de inventario...');
  
  // Usar Redis para evitar que múltiples instancias del servidor corran el cron a la vez
  const lockKey = 'kiora:cron:inventory_check_lock';
  const lock = await redisClient.set(lockKey, 'locked', 'NX', 'EX', 300); // 5 minutos de bloqueo
  
  if (!lock) {
    console.log('[Cron] Otra instancia ya está ejecutando esta tarea. Omitiendo.');
    return;
  }

  try {
    const contextData = await buildSystemContext().catch(() => null);
    const systemPrompt = buildSystemPrompt(contextData);
    const systemMsg = { role: 'system', content: systemPrompt };
    
    // Instrucción para la IA (actuar por iniciativa propia)
    const promptMsg = { 
      role: 'user', 
      content: 'Eres el asistente autónomo del negocio. Acabas de revisar el inventario. Analiza si hay productos críticos (bajo stock) y escribe un mensaje PROACTIVO al administrador informándole. Si encuentras productos bajos, ofrécele enviarle un correo al proveedor (ej. "Tengo X productos bajos, ¿quieres que envíe el correo a Postobon pidiendo Y unidades?"). Si todo está bien, envíale un mensaje de tranquilidad.' 
    };

    const messages = [systemMsg, promptMsg];
    const result = await chatWithTools(messages, TOOLS, executeTool);

    // Enviar a Telegram
    await sendTelegramNotification('🤖 Reporte Autónomo de Inventario', result.response);
    console.log('[Cron] Reporte enviado a Telegram exitosamente.');

  } catch (error) {
    console.error('[Cron] Error en la revisión autónoma:', error.message);
  }
}

function startCronJobs() {
  // Ejecutar todos los días a las 9:00 AM (hora del servidor)
  cron.schedule('0 9 * * *', () => {
    runAutonomousInventoryCheck();
  });
  console.log('[Cron] Tareas programadas iniciadas.');
}

module.exports = { startCronJobs, runAutonomousInventoryCheck };

File: services/ai-service/app/jobs/inventory_cron.py
import logging
from apscheduler.schedulers.asyncio import AsyncIOScheduler
from apscheduler.triggers.cron import CronTrigger
from app.services.telegram_service import send_telegram_notification, redis_client
from app.services.deepseek_service import chat_with_tools
from app.services.system_prompt import build_system_prompt
from app.services.context_builder import build_system_context

logger = logging.getLogger(__name__)

async def run_autonomous_inventory_check():
    logger.info("[Cron] Iniciando revisión autónoma de inventario...")
    
    lock_key = "kiora:cron:inventory_check_lock"
    # Using redis NX to ensure single execution
    lock = await redis_client.set(lock_key, "locked", nx=True, ex=300)
    
    if not lock:
        logger.info("[Cron] Otra instancia ya está ejecutando esta tarea. Omitiendo.")
        return

    try:
        try:
            context_data = await build_system_context()
        except Exception:
            context_data = None
            
        system_prompt = build_system_prompt(context_data)
        system_msg = {"role": "system", "content": system_prompt}
        
        prompt_msg = {
            "role": "user",
            "content": "Eres un agente de compras autónomo. Analiza el inventario para encontrar productos críticos. Por cada producto bajo o agotado, REDACTA UN CORREO formal al proveedor solicitando la cantidad necesaria para reabastecer (basado en la predicción de agotamiento o faltante) y UTILIZA la herramienta 'send_supplier_email' para guardarlo en la carpeta de Borradores de Gmail. Al final, preséntame un resumen que diga: 'He analizado el inventario y guardé X correos en tus Borradores listos para enviar' y dime para qué proveedores fueron. Si todo está bien, envíame un mensaje de tranquilidad."
        }

        messages = [system_msg, prompt_msg]
        result = await chat_with_tools(messages)

        await send_telegram_notification('🤖 Reporte Autónomo de Inventario', result["response"])
        logger.info("[Cron] Reporte enviado a Telegram exitosamente.")

    except Exception as e:
        logger.error(f"[Cron] Error en la revisión autónoma: {e}")

def start_cron_jobs():
    scheduler = AsyncIOScheduler()
    # Ejecutar todos los días a las 3:00 AM (madrugada)
    scheduler.add_job(
        run_autonomous_inventory_check,
        trigger=CronTrigger(hour=3, minute=0)
    )
    scheduler.start()
    logger.info("[Cron] Tareas programadas iniciadas.")
